src/LobbyServer.hs

```diff
@@ -303,3 +295,86 @@ findClient clientName clientList = find ((clientName ==).name) clientList
+
+  cs <- liftIO $ CC.readMVar concurrentChatList
+  if isNothing $ chatName `lookup` cs
+    then do
+      newChatChannel <- liftIO CC.newChan
+      liftIO $ CC.modifyMVar_ concurrentChatList $ \chatList ->
+        return $ (chatName, newChatChannel) : chatList
+      addChannelToClient sid concurrentChatList concurrentClientList
+    else
+      addChannelToClient sid concurrentChatList concurrentClientList
+
+
+    where
+      addChannelToClient sid concurrentChatList concurrentClientList = do
+        cs <- liftIO $ CC.readMVar concurrentChatList
+        let chan = fromJust $ chatName `lookup` cs
+        clientChan <- liftIO $ CC.dupChan chan
+        let newChat = (chatName, clientChan)
+        liftIO $ CC.modifyMVar_ concurrentClientList $ \clients -> do
+          let client = fromJust $ lookupClientEntry sid clients
+          return $ updateListElem (addChatToClient newChat) ((sid ==) . sessionID) clients
+
+      addChatToClient :: Chat -> ClientEntry -> ClientEntry
+      addChatToClient chat client | chatName `elem` map fst (chats client) = client
+                       | otherwise = client {chats = chat : chats client}
+
+-- |Called by a client to read its various chat channels
+readChatChannel :: Server ConcurrentClientList -> String ->  Server ChatMessage
+readChatChannel remoteClientList chatName = do
+  sid <- getSessionID
+  concurrentClientList <- remoteClientList
+  clients <- liftIO $ CC.readMVar concurrentClientList
+  let client = fromJust $ sid `lookupClientEntry` clients
+
+  case chatName `lookup` (chats client) of
+    Nothing -> error $ "LobbyServer.readChatChannel: client hasn't joined chat: " ++ chatName
+    Just chatChannel -> liftIO $ CC.readChan chatChannel
+
+-- | Called by the client to send a chat message
+sendChatMessage :: Server ConcurrentClientList -> Server ConcurrentChatList -> String -> ChatMessage -> Server ()
+sendChatMessage remoteClientList remoteChatList chatName chatMessage = do
+  sid <- getSessionID
+  concurrentClientList <- remoteClientList
+  concurrentChatList <- remoteChatList
+  clientList <- liftIO $ CC.readMVar concurrentClientList
+  chatList <- liftIO $ CC.readMVar concurrentChatList
+
+  case chatName `lookup` chatList of
+    Nothing -> return ()
+    Just chat ->
+      case chatMessage of
+        (ChatMessage from content) -> do
+          msg <- setFromName clientList chatMessage
+          liftIO $ CC.writeChan chat msg
+        ChatJoin          -> do
+          msg <- setFromName clientList $ ChatAnnounceJoin "unknown"
+          liftIO $ CC.writeChan chat msg
+        ChatLeave         -> do
+          msg <- setFromName clientList $ ChatAnnounceLeave "unknown"
+          liftIO $ CC.writeChan chat msg
+  return ()
+    where
+      setFromName :: [ClientEntry] -> ChatMessage -> Server ChatMessage
+      setFromName clientList chatMessage = do
+        sid <- getSessionID
+        let client = fromJust $ lookupClientEntry sid clientList
```

Comment:
Is this a safe use of fromJust? What if the client starts this communication and then immediately disconnects. could crash the server?